### DISCUSSION TITLE
How to upgrade the synaptics vmm5320?

### DISCUSSION THREAD

**Discussion by rokorec31** [fwupd/fwupd#6085] [Q&A] [answered]
I noticed that the builtin.quirk in fwupd has `VMM5331 in Dell dock` configuration, can VMM5320 in Google Gingerbread follow this configuration but use synaptics_mst plugin to support firmware updates?

**Comment by hughsie** [comment]
Using git main branch, what does `sudo fwupdtool -vv get-devices --plugins synaptics_mst` say?

**Reply by rokorec31** [reply]
The latest main branch I build on circleci didn't work on windows, so I installed fedora 38 on VMware to run fwupd. Here is the output.
[get-devices.log](https://github.com/fwupd/fwupd/files/12342952/get-devices.log)

**Reply by hughsie** [reply]
> SMBIOS Manufacturer=VMware, Inc

I don't think MST is going to work in a VM, you need to install Fedora 38 bare metal so to speak.

**Reply by rokorec31** [reply]
Can logs captured by macOS be analyzed?
[get-devices_macOS.pdf](https://github.com/fwupd/fwupd/files/12354251/get-devices_macOS.pdf)

**Reply by superm1** [reply]
It needs to be a real Linux to upgrade DP AUX base device right now. The plugin uses a character device exported by the kernel.

No such abstraction exists in Windows or Mac fwupd today .

**Reply by rokorec31** [reply]
OK, I got it. Here's the log on Fedora 38.
[get-devices_fedora.log](https://github.com/fwupd/fwupd/files/12354629/get-devices_fedora.log)

**Reply by hughsie** [reply]
What commit is that log run with please? i.e. what which commit does `git show` say? I'm wondering if you have a build with 4d83b475d0d5c4ac24bb7b0289f0266905a1681e included?

**Reply by rokorec31** [reply]
What I built was commit `dc00fd0`, should I try `4d83b47` or `6fb2573` instead?

**Reply by hughsie** [reply]
dc00fd0 is fine. This stands out in your log:

    Waiting…                 [*****************                      ]03:18:15.874 FuDevice             using d762543f8c20f636e6fff031a000078d3e10c600 for PCI_SLOT_NAME=0000:00:02.0:drm_dp_aux1
    03:18:15.882 FuDeviceLocker       ignoring close error on aborted open: failed to disable remote control: failed to write command: failed to lseek to 0x4b2 on layer:0, rad:0x0
    03:18:15.882 FuPluginSynapticsMST no cascade device found: failed to enable remote control: failure writing data register: failed to lseek to 0x4c0 on layer:0, rad:0x0
    03:18:15.883 FuDeviceLocker       ignoring close error on aborted open: failed to disable remote control: failed to write command: failed to lseek to 0x4b2 on layer:0, rad:0x1
    03:18:15.883 FuPluginSynapticsMST no cascade device found: failed to enable remote control: failure writing data register: failed to lseek to 0x4c0 on layer:0, rad:0x1
    03:18:15.883 FuDeviceLocker       ignoring close error on aborted open: failed to disable remote control: failed to write command: failed to lseek to 0x4b2 on layer:0, rad:0x2
    03:18:15.883 FuPluginSynapticsMST no cascade device found: failed to enable remote control: failure writing data register: failed to lseek to 0x4c0 on layer:0, rad:0x2
    03:18:15.883 FuPluginSynapticsMST no device found on drm_dp_aux1: ignoring MST-2 device with no SynapticsMstDeviceKind quirk

Does the Synaptics VMM tool work on Windows with this hardware?

**Reply by rokorec31** [reply]
Yes, the VMM tool can works with.
![vmmtool](https://github.com/fwupd/fwupd/assets/10734560/9b7342c2-744b-4626-9563-0accc7c07195)

**Reply by hughsie** [reply]
Do Synaptics have any ideas? I don't understand the MST stuff super well either.

**Reply by superm1** [reply, marked as answer]
> 03:18:15.883 FuPluginSynapticsMST no device found on drm_dp_aux1: ignoring MST-2 device with no SynapticsMstDeviceKind quirk

The problem is `synaptics-mst.quirk` isn't populated for this device AFAICT.

**Reply by rokorec31** [reply]
Thanks, I have get the VMM5320 after added the quirk.
![vmm5320](https://github.com/fwupd/fwupd/assets/10734560/0833f5b4-07e1-4c6a-88e4-055866e884bc)

**Reply by hughsie** [reply]
Right, I think the customer ID offset needs fixing. I've pinged synaptics about this.

**Comment by rokorec31** [comment]
I try  to update vmm5430, it runs out of memory, then freezes without any error message until the process is killed by the OS, do you have any ideas?

fedora: 38
fwupdtool: 1.9.4

![vmm5430 update failure](https://github.com/fwupd/fwupd/assets/10734560/69e1a92b-0c14-4f2d-98e5-9b8bfe05b951)

**Comment by hughsie** [comment]
I've got a branch that fixes quite a few issues -- can you try this please https://github.com/fwupd/fwupd/pull/6110

**Reply by rokorec31** [reply]
I tried #6110 and the freeze gone, but when the update was nearing complete, I got another error message `raw timing data was { 0.000, 13.443, 15.323, 0.001 } -- steps were set as [ 0 98 0 2 ] but should have been [ 0 47 53 0 ] at ../plugins/synaptics-mst/fu-synaptics-mst-device.c:1490 [device-busy]` and the update was not success.

hese's the log: 
[vmm5430 update.log](https://github.com/fwupd/fwupd/files/12427040/vmm5430.update.log)

**Comment by hughsie** [comment]
> I got another error message raw timing data was { 0.000, 13.443, 15.323, 0.001 } -- steps were set as [ 0 98 0 2 ] but should have been [ 0 47 53 0 ] at ../plugins/synaptics-mst/fu-synaptics-mst-device.c:1490 [device-busy]

We can fix that by using the right values and removing GUESSED when the update completes successfully..

>  and the update was not success.

According to the log, the ActiveBank didn't switch -- i.e. you're running the same copy of the firmware as before. The checksum must have matched otherwise we would have failed earlier.  It's possible I broke something in the last commit.

Can you try again with commit 3cf2d90542e48155e520c2440eee328f7ff63acd and also attach the log even if failing? Thanks.

**Reply by rokorec31** [reply]
Sure, here's the log with commit 3cf2d90: 
[vmm5430 update-2.log](https://github.com/fwupd/fwupd/files/12427443/vmm5430.update-2.log)

**Reply by hughsie** [reply]
Hmm, so that last commit didn't seem to break things; the `tag date` output is identical. So I think we're writing the firmware correctly to bank 1, correctly marking bank 1 as valid -- so maybe we're not marking bank 0 as invalid. Can you try with 9d4ce683f92babfa1eaab44fe637b92a8e9060ec perhaps?

**Reply by rokorec31** [reply]
same, here's the log:
[vmm5430 update_9d4ce68.log](https://github.com/fwupd/fwupd/files/12427765/vmm5430.update_9d4ce68.log)

**Reply by hughsie** [reply]
Hmm, that's disappointing. If you undo all my follow up patches and use c726353eea3bc269a30c6a0ad17315c695987b19 does that still fail? If it fails, then maybe I've not broken anything new and we need to ask Synaptics. The c726353eea3bc269a30c6a0ad17315c695987b19 fix should fix the mad OOM.

**Reply by rokorec31** [reply]
c726353 worked, the update succeed.
[vmm5430 update_c726353.log](https://github.com/fwupd/fwupd/files/12428029/vmm5430.update_c726353.log)

**Reply by hughsie** [reply]
Okay, so that's good? news? :) Could you try each commit in turn and tell me what broke things and I can fix it up.

**Reply by rokorec31** [reply]
Yes, but maybe tomorrow later, I'll give you which commit broken.

**Reply by hughsie** [reply]
I managed to reproduce it here when testing the Lenovo dock -- it was a7b1bd61a9a3768a55e8f0a414e73d93441c5def that regressed. I've fixed it up now and force pushed the branch.

**Reply by rokorec31** [reply]
Thanks, the update succeed on Gingerbread.

By the way, if we want to set the quirk setting of vmm5320 in Gingerbread into builtin.quirk as default, what should we do? Should we provide you the quirk settings? or?

**Reply by hughsie** [reply]
Can you do a pull request adding it to one of the existing synaptics-mst .quirk files please.
